src/EditorFeatures/VisualBasic/LineCommit/CommitBufferManager.vb

```diff
@@ -97,63 +96,118 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.LineCommit
-                        ' Avoid using semantics for formatting extremely large dirty spans without an explicit request
-                        ' from the user. The "large span threshold" is 7000 lines. The 7000 line threshold is an
-                        ' estimated value accounting for a lower-bound of the algorithmic complexity of text
-                        ' differencing in designer cases along with measurements of a pathological example demonstrated
-                        ' at 14000 lines. We expect Windows Forms designer formatting operations to run in under ~15
-                        ' seconds on average current hardware when nearing the threshold.
-                        Dim startLineNumber = 0
-                        Dim startCharIndex = 0
-                        Dim endLineNumber = 0
-                        Dim endCharIndex = 0
-                        info.SpanToFormat.GetLinesAndCharacters(startLineNumber, startCharIndex, endLineNumber, endCharIndex)
-                        If endLineNumber - startLineNumber > 7000 Then
-                            useSemantics = False
-                        End If
-                    End If
+            Return CommitDirtyWorkerAsync(isExplicitFormat, blocking, cancellationToken)
+        End Function
 
-                    Dim tree = _dirtyState.BaseDocument.GetSyntaxTreeSynchronously(cancellationToken)
-                    _commitFormatter.CommitRegion(info.SpanToFormat, isExplicitFormat, useSemantics, dirtyRegion, _dirtyState.BaseSnapshot, tree, cancellationToken)
+        Private Async Function CommitDirtyWorkerAsync(
+                isExplicitFormat As Boolean,
+                blocking As Boolean,
+                cancellationToken As CancellationToken) As Task
+            Me.AssertIsForeground()
+
+            ' Create our own linked token to do the async committing work.  This way we can cancel the work if we see
+            ' any edits come through to the buffer.
+
+            Using bufferEditTokenSource = New CancellationTokenSource()
+                Using linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(bufferEditTokenSource.Token, cancellationToken)
+
+                    ' if the text buffer is changed, then cancel any work we're doing.
+                    Dim textBuffer = _dirtyState.BaseSnapshot.TextBuffer
+                    Dim handler = Sub(sender As Object, eventArgs As TextContentChangingEventArgs)
+                                      bufferEditTokenSource.Cancel()
+                                  End Sub
+                    AddHandler textBuffer.Changing, handler
+
+                    Try
+                        ' Start to suppress commits to ensure we don't have any sort of re-entrancy in this process.
+                        ' We've seen bugs (17015) where waits triggered by some computation might re-enter.
+                        Using BeginSuppressingCommits()
```

Comment:
I admit I'm not sure how async plays well with this mechanism. I admit I don't even quite remember what this mechanism was for.